lib/spack/spack/schema/bootstrap.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Schema for bootstrap.yaml configuration file."""
+
+properties = {
+    'bootstrap': {
+        'type': 'object',
+        'properties': {
+            'enable': {'type': 'boolean'},
```

Comment:
I think we should make this a list of enum values so that we can extend it later

```
bootstrap:
  enable: []
```
```
  enable: [source]
```
```
  enable: [binary]
```
```
  enable: [binary, source]
```

Reply:
I considered this design at the beginning and I opted for a Boolean variant. One reason for that is to keep things simple and have a workflow for bootstrapping that always privileges binaries (if there's a trusted source) over sources i.e. I wanted to avoid having to support:
```yaml
bootstrap:
  enable: ['source', 'binary']
```
Another reason is that `binary` isn't a single "entity", but needs to be expanded into a list itself (e.g. Github Actions, LLNL build farm, custom mirrors, ... ). So I was thinking to keep a single list somewhere in this configuration with more structured entries, like:
```yaml
bootstrap:
  binary-providers:
  - name: github-actions
    description: 'Binary mirrors built on the cloud'
    type: buildcache
    trusted: unknown
  - name: llnl
    description: 'Binary mirrors from the LLNL build farm'
    type: buildcache
    trusted: unknown
```
So far I didn't consider disabling bootstrapping from sources as a last resort, but we can add that if that's a valid use case.

This is to say that I'm obviously open to discussing this, but would it be fine with you to merge as is for the time being and have the discussion when we'll move to #22720 and add also the trust mechanism for binaries? I think in that PR we'll be in a better position to try different solutions if need be.

Reply:
I'm worried that if we merge this, we'll be stuck with backwards compatibility code (or an `update` method for `bootstrap.yaml`) when we change it later. I hadn't actually considered the issue of `enable: [source, binary]` though, I think I was considering the entries to be a set and Spack would always prefer binaries if it's an option, and fall back to source if there was no binary trusted option for that particular dependency. So I was thinking something like

```
bootstrap:
  enable: [source, binary]
  binary_providers:
  - name: mymirror
    description: ...
    type: # I'm not sure what this does
    trusted: true
  - name: spack-llnl
    ...
    trusted: false
  - name: spack-github
    ...
    trusted: unknown
```
Would mean for each dependency.
1. prompt whether we should trust `spack-github` (assume they say yes for this example)
2. Check for it in mymirror, if it's not there or install fails then
3. Check for it in spack-github, if it's not there or install fails then
4. Build from source

Reply:
Yeah, what I would do there is something more like:
```yaml
bootstrap:
  enable: true  # This just means I want to try bootstrapping
  providers: # Here I have the methods in order with enough info to use them
  - name: mymirror
    description: ...
    type: buildcache  # This one bootstraps from a Spack buildcache
    trusted: true
```
If we want to make bootstrapping from source explicit we can add an entry like:
```yaml
- name: spack-build
  description: "From source build of the missing dependency"
  type: source_build
  trusted: true
```
The entry `type` is meant to identify the kind of source we are using. At the moment we have only `buildcache` and `source_build` (names are just placeholders, of course) but we may figure out other ways later e.g. install something on the fly with `pip` in a virtual env.

Basically the idea is to use enable to check whether we want to do bootstrap or not. Then use a list of objects to express the order in which we want to try things.

Reply:
> I'm worried that if we merge this, we'll be stuck with backwards compatibility code (or an update method for bootstrap.yaml) when we change it later.

I'd say if it is not in a tagged release it never existed :laughing: Getting more serious, if there are no other things to change in this PR than the schema for `enable` I wonder if we can:
1. Rebase the #22720 branch on this one
2. Have another PR from that branch to this branch 
 
and still have CI running (since we are not targeting `develop`). If we can do that we can hold on on merging this to develop, and work in another PR whether we should go in one direction or the other since we have also bootstrapping from binaries available. What do you think?

Reply:
I think it's fine as long as we don't have to write backwards compatibility code for it just because it's in develop for a little while.